### DISCUSSION TITLE
Setting Black Point in RGB channels

### DISCUSSION THREAD

**Discussion by retyazhi** [ImageMagick/ImageMagick#7620] [Help]
### ImageMagick version

7.1.0-27

### Operating system, version and so on

Windows 10

### Description

Hello!

Is it possible in Imagemagick to set an exact black point of an image in RGB channels? For example, to set the black point for Red channel at 10, for Green channel at 15 and for Blue channel at 22? I need it for correcting my film scans.

**Comment by snibgo** [comment]
We can adjust an image so that the red channel from 10 to 255 (on a scale of 0 to 255) becomes 0 to 255, and similarly for green and blue. Windows BAT script:
```
magick ^
  in.tiff ^
  -channel red -level %%[fx:10/255*100],100%% ^
  -channel green -level %%[fx:15/255*100],100%% ^
  -channel blue -level %%[fx:22/255*100],100%% ^
  +channel ^
  out.tiff
```
In a direct Windows CMD command, change the double percents `%%` to single percent `%`.

Note that "`%%[fx:10/255*100]`" converts 10 (on a scale of 0 to 255) to a percentage, which is 3.9216%.

See also https://imagemagick.org/script/command-line-options.php#level

Is that what you want?

**Reply by retyazhi** [reply]
Yes, it looks like just what I need, though it would be better to set exact numbers, not in percents. However, I'm getting an error with this code. Maybe I'm doing something wrong? Here is the error (the same error for all RGB channels):

invalid argument for option '-level': %[fx:10/255*100],100% @ error/convert.c/ConvertImageCommand/2038.

**Comment by snibgo** [comment]
I should have included `+channel`. I have edited the above.

Are you running this in a BAT script? If so, then it should work.

But if you are running this from a terminal (aka "DOS window" or "console window" aka CMD interface), then change the double percents `%%` to single percent `%`.

Yes, you can use exact numbers instead of percentages. But the numbers will be on a scale of 0 to QuantumRange. QuantumRange is either 255 if your IM is Q8, or 65535 if your IM is Q16, or 4294967295 if your IM is Q32.

The command `magick -version` will show the Q-number on the first line.

For example, if your IM is Q16 then try this (Windows BAT syntax):
```
magick ^
  in.tiff ^
  -channel red -level 2570,%%[fx:QuantumRange] ^
  -channel green -level 3855,%%[fx:QuantumRange] ^
  -channel blue -level 5654,%%[fx:QuantumRange] ^
  +channel ^
  out.tiff
```
Where did 2570 etc come from? You wanted 10 (out of 255). For a scale of 0 to 65535, we need to divide 10 by 255 and multiply by 65535. `10*65535/255 = 2570`.

**Reply by retyazhi** [reply]
I'm still having difficulties. I'm running the CMD interface. My IM is Q16. If we just stick with one channel, do I understand correctly that the needed part of the code will look like
-channel red -level 2570,%[fx:QuantumRange] +channel

I've also tried
-channel red -level 2570,%[fx:QuantumRange] ^ +channel ^

but I'm getting errors with both of them. I also tried to change 'QuantumRange' with 64435 but it din't help either :)

Could you, please, write this part of the code for an CMD interface?

**Reply by fmw42** [reply]
snibgo: you wrote 

> 64435 if your IM is Q16

but should that not be 

`65535 if your IM is Q16`

**Reply by snibgo** [reply]
Thanks for the correction.

**Comment by snibgo** [comment]
A simple command, for CMD, is:
```
magick in.png -channel red -level 2570,%[fx:QuantumRange] +channel out.png
```
The larger command above is split into lines. The caret ^ at the end of each line means the command is continued on the next line. When you have a single-line command, it doesn't need carets.

> but I'm getting errors with both of them.

When you get errors, please paste the entire error line here.

**Comment by retyazhi** [comment]
Now it works! Thank you so much, snibgo!

Could you, please, also tell me how can I use Imagemagick scripts (http://www.fmwconcepts.com/imagemagick/index.php)? Do I understand corectly that entering codes with the CMD is not an efficient way of using Imagemagic potetial?

**Comment by snibgo** [comment]
Fred is a user on this forum (fmw42). His scripts are for bash, which is a shell. (Windows CMD is also a shell.) On my computer I use Cygwin bash. An alternative is WSL.

On Fred's page that you linked, see the section starting "Windows users can run my scripts, if they install the unix Cygwin system for Windows and then install the IM Cygwin binary."

> Do I understand corectly that entering codes with the CMD is not an efficient way of using Imagemagic potetial?

I don't know what you mean.

**Reply by retyazhi** [reply]
Thank you once again for your help!!
